lib/nat/registry.go

```diff
@@ -19,34 +19,33 @@ func Register(provider DiscoverFunc) {
 	providers = append(providers, provider)
 }
 
-func discoverAll(renewal, timeout time.Duration) map[string]Device {
+func discoverAll(renewal, timeout time.Duration, stop chan struct{}) map[string]Device {
 	wg := &sync.WaitGroup{}
 	wg.Add(len(providers))
 
 	c := make(chan Device)
-	done := make(chan struct{})
-
 	for _, discoverFunc := range providers {
 		go func(f DiscoverFunc) {
 			for _, dev := range f(renewal, timeout) {
-				c <- dev
+				select {
+				case c <- dev:
+				case <-stop:
```

Comment:
I'd add an explicit return to avoid having to understand how long time the `f(...)` call might take for each device while exiting

Reply:
This will leave hanging routines tho, we should push the stop down and pass in a context. The wg.Done should be moved up too.

Reply:
The changes were anyway wrong (my assumption of number of elements in `nats` was incorrect). I reverted everything except the stop-select part.